design/ingressroute-v1-design.md

```diff
@@ -0,0 +1,320 @@
+- `match` previously took a string argument which was the pathPrefix to route on, now this has been moved to a `path` parameter under the `match` struct
+- `tcpproxy`: Is removed and replaced by a `supported protoco` defined in the route
+- Annotations on Kubernetes services are removed which previously defined what ports to proxy TLS connections to, etc
+- **TODO**: Investigate moving from `spec.status` to the built-in [status sub resource type](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#status-subresource).
+- **retryPolicy** renamed to `failurePolicy`
+
+### Routes
+
+Routes in v1beta1 allows for a set of matches on a path as well as a set of upstream Kubernetes services that traffic could be routed. These upstreams assumed an l7 http protocol+port unless [special annotations](https://github.com/heptio/contour/blob/master/docs/annotations.md#contour-specific-service-annotations) were added to the corresponding Kubernetes service. These mappings needed to be tightly coupled and were often overlooked by users.
+
+This design looks to update how services are defined by specifying the protocol in the service reference set. By defining the protocol:service name as well as the port, we can eliminate the need for annotations on the service.
+
+<u>Supported Proctocols:</u>
+
+- **http**: l7 insecure http proxy
+- **https**: l7 secure (tls) http proxy
+- **http2**: l7 insecure http2 proxy
+- **https2**: l7 secure http2 proxy
+- **tcp**: TLS encapsulated TCP sessions
+
+#### Example:
+
+```yaml
+routes: 
+  - match: /
+    backends: 
+    - http: kuard
+      port: 80
+    - https: kuard-tls
+      port: 443
+    - http2: kuard-2
+      port: 80
+    - https2: kuard-2-tls
+      port: 443
+```
+
+**NOTE**: *Implementing specific protocols in the services section eliminates the `tcpproxy` field which previously implemented tcp proxying*
+
+## Header Routing
+
+The IngressRoute spec will be updated to allow for a `header` field to be added which will allow for a set of key/value pairs to be applied to the route. Additionally, the path `match` moves to it's own `path` variable.
+
+**NOTE**: *Routing on a `header` value* still requires a path prefix to be specified. If pure header delegation is requested, then Contour will configure a wildcard path to match all paths resulting in a header match only across any path.
+
+#### Header Values
+
+IngressRoutev1 will support routing from various headers. It will implement commonly used headers such as user-agent & cookie since those will require a regex style implementation. By implementing them as speific types, we can apply those regex's automatically.
+
+- **header:** This is a generic header type and allows for matching custom headers
+- **cookie**: This is a built-in type and matches a `cookie` in the header while being implemented as a "contains" regex
```

Comment:
Cookies are just headers with a fixed value. Can we simplify cookie routing into a header called cookie and substring matching for key=value?

Reply:
We could, I was just looking for ways to simplify the IngressRoute document. We're adding more cases and tests in Contour, but the user has a better UX. We could also start here and design the other later.

Reply:
> Cookie and user-agent feel like helpers. If we give people flexible enough ways to match on headers generically, they can implement Cookie and User-agent themselves. I think this would be enough for IngressRoute 1 and we can add cookie and user-agent, and perhaps others, when needed, post contour 1.0.

👍